yjit/src/codegen.rs

```diff
@@ -3979,6 +3963,38 @@ fn gen_send_cfunc(
     asm.cmp(CFP, stack_limit);
     asm.jbe(counted_exit!(ocb, side_exit, send_se_cf_overflow).into());
 
+    if flags & VM_CALL_ARGS_SPLAT != 0 {
+        if cfunc_argc > 0 {
+            // push_splat_args does a ctx.stack_push so we can no longer side exit
+            argc = push_splat_args(argc, ctx, asm, ocb, side_exit, cfunc_argc as u32);
+        } else {
+            // This is a variadic c function and we'd need to pop
+            // each of the elements off the array, but the array may be dynamically sized
+            gen_counter_incr!(asm, send_args_splat_variadic);
+            return CantCompile
+        }
+    }
+
+    // Number of args which will be passed through to the callee
+    // This is adjusted by the kwargs being combined into a hash.
+    let passed_argc = if kw_arg.is_null() {
+        argc
+    } else {
+        argc - kw_arg_num + 1
+    };
+
+    // If the argument count doesn't match
+    if cfunc_argc >= 0 && cfunc_argc != passed_argc {
+        gen_counter_incr!(asm, send_cfunc_argc_mismatch);
+        return CantCompile;
+    }
+
+    // Don't JIT functions that need C stack arguments for now
+    if cfunc_argc >= 0 && passed_argc + 1 > (C_ARG_OPNDS.len() as i32) {
+        gen_counter_incr!(asm, send_cfunc_toomany_args);
+        return CantCompile;
+    }
+
```

Comment:
This crashes reliably on x86_64 w/ gcc (at least on my arch linux machine)

I suspect because

> // push_splat_args does a ctx.stack_push so we can no longer side exit

but then the `return CantCompile` lower down _are_ side exits.

Here's a standalone reproduction which may help:

``` ruby
require "psych"

@emitter = Psych::Emitter.new @out

[
  [:foo, nil, nil, false, true, 1]
].each do |args|
  @emitter.scalar(*args)
end
```

run as:
```
$ make -j ruby && ./ruby --yjit-call-threshold=1 test_crash_args_splat.rb
test_crash_args_splat.rb:10: [BUG] Segmentation fault at 0x0000000000000003
ruby 3.2.0dev (2022-09-14T22:57:39Z yjit_push_frame 35e86edccc) +YJIT [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0004 p:0009 s:0020 e:000013 BLOCK  test_crash_args_splat.rb:10 [FINISH]
c:0003 p:---- s:0010 e:000009 CFUNC  :each
c:0002 p:0023 s:0006 e:000005 EVAL   test_crash_args_splat.rb:9 [FINISH]
c:0001 p:0000 s:0003 E:0009e0 (none) [FINISH]
```

Reply:
Thanks for finding this. Will take a look

Reply:
merged https://github.com/ruby/ruby/pull/6385. thanks Jimmy!